test/session.test.js

```diff
@@ -0,0 +1,160 @@
+/* eslint-disable no-unused-expressions */
+/* eslint-disable no-undef */
+import * as Factory from './lib/factory.js';
+chai.use(chaiAsPromised);
+const expect = chai.expect;
+
+describe('server session', () => {
+  const BASE_ITEM_COUNT = 1; /** Default items key */
+
+  const syncOptions = {
+    checkIntegrity: true,
+    awaitAll: true
+  };
+
+  before(async function () {
+    localStorage.clear();
+  });
+
+  after(async function () {
+    localStorage.clear();
+  });
+
+  beforeEach(async function () {
+    this.timeout(Factory.TestTimeout);
+    this.expectedItemCount = BASE_ITEM_COUNT;
+    this.application = await Factory.createInitAppWithRandNamespace();
+    this.email = Uuid.GenerateUuidSynchronously();
+    this.password = Uuid.GenerateUuidSynchronously();
+    await Factory.registerUserToApplication({
+      application: this.application,
+      email: this.email,
+      password: this.password
+    });
+  });
+
+  afterEach(async function () {
+    expect(this.application.syncService.isOutOfSync()).to.equal(false);
+    const items = this.application.itemManager.items;
+    expect(items.length).to.equal(this.expectedItemCount);
+    const rawPayloads = await this.application.storageService.getAllRawPayloads();
+    expect(rawPayloads.length).to.equal(this.expectedItemCount);
+    await this.application.deinit();
+  });
+
+  function getDelayBeforeNextRequest(currentSession) {
+    const tokenExpireAt = currentSession.expireAt;
+    const timeRemaining = Math.ceil(tokenExpireAt - Date.now() / 1000);
+    /* 
+      If the token has not expired yet, we will return the remaining time, plus one second.
+      Else, there's no need to add a delay.
+    */
+    return timeRemaining > 0 ? timeRemaining + 1 : 0;
+  }
+
+  async function getSessionFromStorage(application) {
+    return application.storageService.getValue(StorageKey.Session);
+  }
+
+  it('should fail when a sync request is perfomed with an expired access token', async function () {
+    const currentSession = this.application.apiService.session;
+
+    const delayBeforeNextRequest = getDelayBeforeNextRequest(currentSession);
+    await Factory.sleep(delayBeforeNextRequest);
+
+    const response = await this.application.apiService.sync([]);
```

Comment:
What's the empty array for? `sync` doesn't take an array param.

Reply:
The empty array is for the `payloads` param: https://github.com/standardnotes/snjs/blob/fbc5ec68ad47c47eaccf67ceb7971b0b43f8b36a/lib/services/api/api_service.ts#L231

If the `payloads` param is not provided, [this line](https://github.com/standardnotes/snjs/blob/fbc5ec68ad47c47eaccf67ceb7971b0b43f8b36a/lib/services/api/api_service.ts#L244) will fail because of the `map` method running on an undefined variable.

Also, I'm using `this.application.apiService.sync([])` and not `this.application.syncService.sync()` because I want to have the response available.

Reply:
Ohh I didn't catch that this was the apiService and not syncService.